### PR TITLE
Feature/issues/526

### PR DESCRIPTION
**Pull request by Mizzick**
#526 FF Performance issues

### CODE REVIEW

File: Extension/browser/firefox/lib/contentPolicy.js [hunk 1]
```diff
@@ -143,15 +143,30 @@ var WebRequestHelper = exports.WebRequestHelper = {
...
-            if ("" + context === "[object ChromeWindow]") {
-                return null;
-            }
             return tabUtils.getTabForContentWindow(context.top);
         }
+
         return null;
     },
 
+    getDocumentContextWindow: function (context) {
```
Comment (atropnikov) [inline Extension/browser/firefox/lib/contentPolicy.js]: Unused

File: Extension/browser/firefox/lib/contentPolicy.js [hunk 2]
```diff
@@ -166,18 +181,20 @@ var WebRequestHelper = exports.WebRequestHelper = {
         if (contextData && contextData.browser) {
             var browser = contextData.browser;
 
-            var xulTab = tabUtils.getTabForBrowser(browser);
+            if (typeof tabUtils.getTabForBrowser === 'function') {
```
Comment (atropnikov) [inline Extension/browser/firefox/lib/contentPolicy.js]: `tabUtils` is module, it always has `getTabForBrowser` function

File: Extension/lib/filter/filters.js [hunk 3]
```diff
@@ -433,7 +438,12 @@ RequestFilter.prototype = {
...
             this._clearRequestCache();
         }
-        this.requestCache[requestUrl] = [rule, refHost, requestType];
+
+        if (!this.requestCache[requestUrl]) {
+            this.requestCache[requestUrl] = Object.create(null);
+        }
+
+        this.requestCache[requestUrl][requestType] = [rule, refHost];
         this.requestCacheSize++;
```
Comment (atropnikov) [inline Extension/lib/filter/filters.js]: You don't need to increment cache size each time. You've already known that key is in the cache or not.

File: Extension/browser/firefox/lib/contentScripts.js [hunk 4]
```diff
@@ -318,11 +318,17 @@ ContentScripts.prototype = {
         var listener = (function (contentMessageHandler) {
 
             function getTabFromTarget(target) {
-                var tab = tabUtils.getTabForBrowser(target);
+                var tab;
+
+                if (typeof tabUtils.getTabForBrowser === 'function') {
```
Comment (atropnikov) [inline Extension/browser/firefox/lib/contentScripts.js]: See comment above

File: Extension/browser/firefox/lib/contentPolicy.js [hunk 5]
```diff
@@ -129,12 +129,12 @@ var WebRequestHelper = exports.WebRequestHelper = {
```
Comment (atropnikov) [inline Extension/browser/firefox/lib/contentPolicy.js]: I think no need to remove these checks (except `if ("" + context === "[object ChromeWindow]")`). You have to create a cache for `getTabForContentWindow` function result. Something like `cache = WeakMap(winTop, tab)`

File: Extension/browser/firefox/lib/contentPolicy.js (hunk 2 above)
Comment (Mizzick) [inline Extension/browser/firefox/lib/contentPolicy.js]: But unexpectively I've seen these errors in console.

File: Extension/browser/firefox/lib/contentPolicy.js (hunk 5 above)
Comment (Mizzick) [inline Extension/browser/firefox/lib/contentPolicy.js]: These checks eat a lot of resources, and there is no need in it actually.

File: Extension/lib/filter/filters.js (hunk 3 above)
Comment (Mizzick) [inline Extension/lib/filter/filters.js]: this incrementation is for cleaning.

File: Extension/browser/firefox/lib/contentPolicy.js (hunk 5 above)
Comment (Mizzick) [inline Extension/browser/firefox/lib/contentPolicy.js]: And I'm not sure that we can make a cache here, cause the context parameter could be different each time.

File: Extension/lib/filter/filters.js (hunk 3 above)
Comment (atropnikov) [inline Extension/lib/filter/filters.js]: I know, but in FF you receive many requests with the same URL. It leads to cache clearing more often than expected.

File: Extension/browser/firefox/lib/contentPolicy.js (hunk 2 above)
Comment (atropnikov) [inline Extension/browser/firefox/lib/contentPolicy.js]: So your fix doesn't resolve the problem.

Comment (Mizzick) [inline Extension/browser/firefox/lib/contentPolicy.js]: No, my fix resolved the problem with those errors in console.

File: Extension/lib/filter/filters.js (hunk 3 above)
Comment (Mizzick) [inline Extension/lib/filter/filters.js]: Ah, ok

**Pull request merged**